draft-ietf-ppm-dap.md

```diff
@@ -1514,217 +1614,253 @@ to the next phase of the aggregation protocol.
+and the Helper stores its output shares for use in the collection flow
+({{collect-flow}}). If `state == Continued(round, prep_state)` for each report,
+then more rounds of preparation are required and the Helper proceeds as
+described below. (Note: Since VDAF preparation completes in a constant number
+of rounds, it will never be the case that some reports are completed and others
+are not.)
+
+the Helper awaits an HTTP POST request to the aggregation job URI from the
+Leader, the body of which is an `AggregationJobContinueReq` as specified in
+{{aggregation-leader-continuation}}.
+
+Next, Helper begins by checking that it recognizes the task ID. If not, then it
+MUST abort with error `unrecognizedTask`.
 
-Otherwise, the Helper continues with preparation for a report share by combining
-the previous message round's prepare message (carried by the AggregationJobReq)
-and its current preparation state (`prep_state`). This step yields one of three
-outputs:
+Next, it checks if it recognizes the indicated aggregation job ID. If not, it
+MUST abort with error `unrecognizedAggregationJob`.
 
-1. An error, in which case the report share is marked as invalid.
-1. An output share, in which case the helper stores the output share for future
-   collection as described in {{collect-flow}}.
-1. Its next VDAF preparation state and message-share, denoted `(next_prep_state,
-   next_prep_share)`.
+Next, the Helper checks that:
 
-To carry out this step, for each `PrepareStep` in `AggregationJob.prepare_steps`
-received from the leader, the helper performs the following check to determine
-if it should continue preparing the report share:
+1. the report IDs are all distinct
+1. each report ID corresponds to one of the `state` objects
+1. each prepare step has message type "continue"
+1. `AggregationJobContinueReq.step` is not equal to `0`
 
-* If the status is `failed`, then mark the report as failed and reply with a
-  failed PrepareStep to the Leader.
-* If the status is `finished`, then mark the report as finished and reply with a
-  finished PrepareStep to the leader. The Helper then stores the output share
-  and awaits collection; see {{collect-flow}}.
+If any of these checks fail, then the Helper MUST abort with error
+`invalidMessage`. Additionally, if any prep step appears out of order relative
+to the previous request, then the Helper MAY abort with error `invalidMessage`.
+(Note that a report may be missing, in which case the Helper should assume the
+Leader rejected it.)
 
-Otherwise, preparation continues. The Helper MUST check its current round
-against the Leader's `AggregationJobContinueReq.round` value. If the Leader is
-one round ahead of the Helper, then the Helper combines the Leader's prepare
-message and the Helper's current preparation state as follows:
+[OPEN ISSUE: Issue 448: It may be useful for the Leader to explicitly signal
+rejection.]
+
+Next, the Helper checks if the continuation step indicated by the request is
```

Comment:
There are a few questions to me as to whether the step tracking is still needed; I think it is, and this text is why. Without tracking aggregation progress in some way, we wouldn't be able to recognize when we're in a skew state.

Of course, the DAP step is now redundant with the VDAF round, which as of the latest VDAF ping-pong iteration is tracked in the state. So, we could drop the DAP step & track retries with VDAF round. But I think this isn't a good idea for two reasons:

1. The DAP round goes 1, 2, 3...; the VDAF rounds that are recorded will go 2, 4, 6... This would mean that a continuation request could be sent with an odd round; I suppose we should fail in this case, but an implementation will be happier if we arrange things such that this error condition can't exist at all.

2. The VDAF round only needs to be tracked in state because `ping_pong_next` needs to use it to parse the `inbound` message. But this means that parsing the `inbound` message will require shared context (i.e. correct parsing requires the aggregators to agree on the VDAF round). Since the current round information is sent in the same `AggregationContinueReq` as the `inbound` message itself, this means that aggregator implementations will need to ensure they split parsing. If they don't, they'll see rare errors that occur only during round-skew.

Reply:
> There are a few questions to me as to whether the step tracking is still needed; I think it is, and this text is why. Without tracking aggregation progress in some way, we wouldn't be able to recognize when we're in a skew state.

I agree that continuation step tracking is necessary in order to implement skew recovery -- what I'm questioning is whether we need such explicit language about it. For instance we say things like:

> The Leader is now in DAP step 0.
> The Leader advances to the next DAP step.
> The Leader sends a message to the Helper to say that it should advance to the next DAP step.

I think it would be simpler to just say that the first continuation request has `step` set to `1`, the second has `step` set to `0` and so on, and say that the Helper uses this field for step-skew recovery. In fact I don't think any of the paragraph needs to change as long as it's clear what the value of `step` is supposed to be. We might even stick this in the ladder diagram to make sure it's crystal clear.

> Of course, the DAP step is now redundant with the VDAF round, which as of the latest VDAF ping-pong iteration is tracked in the state. So, we could drop the DAP step & track retries with VDAF round. But I think this isn't a good idea for two reasons:

Totally agreed, let's not even bring up the VDAF round number in this section.